### PR TITLE
[Caffe2] Added HIP top_k operator

### PR DESCRIPTION
**Pull request by ashishfarmer**
This PR contains changes for:
1. Adding HIP top_k operator in Caffe2
2. Added HIP equivalent definitions of GPUDefs and GPUScanUtils
3. Removing the top_k operator test from ROCm test ignore list
4. Bug fixes in related code in THC/THCAsmUtils.cuh

cc: @bddppq

### CODE REVIEW

**Comment by ashishfarmer** [comment]
Triggered tests: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/5675/

File: caffe2/utils/GpuScanUtils.cuh [hunk 1]
```diff
@@ -115,7 +122,11 @@ __device__ void exclusiveBinaryPrefixScan(T* smem, bool in, T* out, T* carry, Bi
   *out -= (T) in;
 
   // The outgoing carry for all threads is the last warp's sum
+#if defined(__HIP_PLATFORM_HCC__)
+  *carry = smem[math::divUp<int>(blockDim.x, kWarpSize) - 1];
+#else
   *carry = smem[(blockDim.x / 32) - 1];
```
[the remark this answers is not in the document]
Comment (ashishfarmer) [reply caffe2/utils/GpuScanUtils.cuh]: Missed that. Thanks for catching it

**Comment by ashishfarmer** [comment]
Triggered tests: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/5685/

**Comment by bddppq** [comment]
@ashishfarmer After we have got more amd machines last week, rocm tests are automatically run now :-) (so no need to manually trigger).

**Review by bddppq — APPROVED** [review]
LG. Thanks!

**Pull request closed without merging**